### ISSUE TITLE
Lazy Loading Nonrouteable Engines

### ISSUE THREAD

**Issue by ooluwadiya** [dgeb/ember-engines#232]
I don't think this feature has been implemented / tested but I'll go ahead and log an issue for tracking. Lazy-loading non-routeable engines is not working. The assets are never loaded and as such are not registered...

I end up with this error  "Assertion Failed: You used `{{mount 'nonrouteable'}}`, but the engine 'nonrouteable' can not be found.""

**Comment by trentmwillis** [comment]
Yeah, this hasn't really been considered yet. I think the primary question was how to handle the fact that the mount point will render asynchronously. Once that is determined, we just need to identify where we in that lifecycle we initiate the request to load the Engine assets.

**Comment by ooluwadiya** [comment]
That makes sense. Thanks

**Comment by aalasolutions** [comment]
Any updates on this? Our app is route-less and need to use lazy loading for the engine. Main JS file is like 3MB and some users are not having access to all the areas.

**Comment by villander** [comment]
> Route-less Engines, on the other hand, are Engines which are not able to be routed to. This means that they don’t live at a specific route, but are instead mounted within a template or your application. Examples of good candidates for route-less Engines are chat boxes or complex forms.

Following the concept of route-less, I think we should start the request to load the Engine assets, when the template of that route / component is called. We can do something like what React does:

https://gist.github.com/rubenspgcavalcante/ea8d087cc2d997392abb2f1c7e02b666#file-app-jsx

> Yeah, this hasn't really been considered yet. I think the primary question was how to handle the fact that the mount point will render asynchronously. Once that is determined, we just need to identify where we in that lifecycle we initiate the request to load the Engine assets.

**Comment by GCheung55** [comment]
The recent talk by [Dan Ambramov about async rendering](https://www.youtube.com/watch?v=v6iR3Zk4oDY) made me think of this lazy loading routless engines issue.

**Comment by anulman** [comment]
Bumping this thread; per @villander's React notes I'd also suggest looking upstream at Webpack's code splitting guide, particularly re [prefetch & preload](https://webpack.js.org/guides/code-splitting/#prefetching-preloading-modules).

Since Ember is convention > configuration though, I think our syntax shifts slightly. Specifically, I think non-routable `{{mount 'foo'}}`s ought to:

- Accept an `eager=true` arg as a local override to the application's `lazyLoading: true` config—at build time this engine would be compiled into its host;
- Accept `prefetch=true` to lazy-load the engine once the host has completed loading & booting; and
- Otherwise default to lazy-loading at render time

In all cases, `{{mount}}` should accept actions for lifecycle hooks (eg `willLoad`, `didLoad`, `willRender`, `didRender`, `onError`) and expose load & error states via contextual component (eg `isLoading` & `didError`).

This would look something like:

Lazy-load a chat room on a parent route's template:
```handlebars
{{outlet}}
{{#mount 'chat-room' as |engine|}}
  {{#engine.isLoading}} It never hurts to try<br />{{loading-spinner}} {{/engine.isLoading}}
  {{#engine.didError}} Could not load chat; reload the page or <button onclick={{action engine.retry}}>click to retry</button>
{{/mount}}
```

Prefetch a login modal once host has loaded + booted:
```handlebars
{{#if isShowingLoginModal}}
  {{#mount 'login-modal' prefetch=true didLoad=(action 'instrumentLoginLoad') as |engine|}}
    {{#engine.isLoading}} {{loading-spinner}} {{/engine.isLoading}}
    {{#engine.didError}} Load error; <button onclick={{action engine.retry}}>click to retry</button> {{/engine.didError}}
{{/if}}
```

Opt-in for old behaviour (ie expect the engine to be loaded):
```handlebars
{{#if isShowingLoginModal}}
  {{mount 'login-modal'}}
{{/if}}
```

If useful, I have a spike of another approach to code-splitting modules I'm calling Ember Async Registry, which has a few tradeoffs vs. Ember Engines and IMO is more useful in route-less engine scenarios. Happy to share more if of interest; I've written a draft blog post introducing the concept (for publishing week of Jul 9), but do not yet have an addon any app can just "drop in".

**Comment by toovy** [comment]
I'm curious, is anyone still working on this?

**Comment by rwjblue** [comment]
AFAICT https://github.com/ember-engines/ember-asset-loader/issues/56 is the only thing required...

**Comment by toovy** [comment]
@rwjblue thanks, sounds like it is a simple step. You already posted reference code. What is hindering the implementation?

**Comment by rwjblue** [comment]
Someone to do it 😸

**Comment by toovy** [comment]
Guess then we'll have to wait for someone smart having the time! :)

**Comment by buschtoens** [comment]
Cross-posting https://github.com/ember-engines/ember-asset-loader/issues/56#issuecomment-397867856:

> The most recent Ember.js times newsletter had a reference to this, that reminded me about this addon I once wrote: [`ember-lazy-mount`](https://github.com/buschtoens/ember-lazy-mount)
> 
> I guess this is not really relevant, since it uses a component instead of a template helper, but maybe you still find it useful. ✌️

**Comment by toovy** [comment]
Just tested https://github.com/buschtoens/ember-lazy-mount and it seems to work fine. As I now know this addon now I'm fine with the solution, still I think the code somehow should be incorporated in standard engines (or asset loader). This feature is really great, once more it enables you to manage big applications and split the code apart. A great win for ember.

Shouldn't we update the readme at https://github.com/ember-engines/ember-engines#lazy-loading-engines? I guess lazy loading is fully supported now.

Furthermore the docs at http://ember-engines.com/guide/lazy-loading says that 

> Currently, lazy loading is not supported for route-less Engines. Progress can be tracked at this issue.

With a reference to this issue. Ok now you will find a solution once you look into the issue. But maybe the documentation should directly point to the addon so that the reader does not assume that this feature is not available or "an issue".

What do you think?

**Comment by YukiDayDreamer** [comment]
Hi @toovy , does this addon solution still works for you? I tried to use the lastest ember-cli and ember-engines, but it will throw an error message **_Uncaught TypeError: params is not iterable_**.

**Comment by toovy** [comment]
@YukiDayDreamer sorry, had no chance to try it again, maybe the release of ember 3.6 is the issue?

**Comment by buschtoens** [comment]
@toovy The issue is tracked here: https://github.com/buschtoens/ember-lazy-mount/issues/2. It should be fixed; I am waiting for a response from @YukiDayDreamer . 😊

**Comment by villander** [comment]
@stefanpenner @dgeb @rwjblue I would love to evolve with this, do you want to make a meeting about it?
